File: mobile/plugins/home-widgets/ios/WidgetPayloadContract.swift
import Foundation

let widgetSupportedSchema = 1
let widgetIstTimeZone = "Asia/Kolkata"
let widgetLanguages = ["hi", "en", "gu", "kn"]

struct LocalizedText: Codable {
  let hi: String; let en: String; let gu: String; let kn: String
  func value(_ locale: String) -> String { locale == "en" ? en : locale == "gu" ? gu : locale == "kn" ? kn : hi }
}
struct LocalizedLines: Codable {
  let hi: [String]; let en: [String]; let gu: [String]; let kn: [String]
  func value(_ locale: String) -> [String] { locale == "en" ? en : locale == "gu" ? gu : locale == "kn" ? kn : hi }
}
struct VerseDay: Codable { let dateKey, sourceId: String; let chapter: Int?; let verseIndex: Int; let lines: LocalizedLines; let excerpt, source, accessibilityLabel: LocalizedText; let deepLink: String }
struct VerseSlice: Codable { let timeZone, validThrough: String; let days: [VerseDay] }
/// One link of the day's tithi chain (contract.ts `PanchangWidgetTithi`), in
/// order, link 0 always the sunrise tithi `PanchangDay.tithi` names. Optional
/// end: the last link's end belongs to tomorrow's solve, never a guess.
struct PanchangTithi: Codable { let name: LocalizedText; let endsAt: String?; let till: LocalizedText? }
struct PanchangDay: Codable { let dateKey: String; let representedDate, tithi: LocalizedText; let tithiSegments: [PanchangTithi]?; let vrat: LocalizedText?; let sunrise, rahuKaal: LocalizedText; let abhijit: LocalizedText?; let deepLink: String }
struct PanchangSlice: Codable { let timeZone, cityId: String; let cityLabel: LocalizedText; let calendarSystem, validThrough: String; let days: [PanchangDay] }
struct JapamSlice: Codable { let dateKey, timeZone: String; let totalBeads, totalRounds, japaStreak: Int; let lastUsedMantraId: String?; let deepLink: String }
struct WidgetPayload: Codable { let schemaVersion: Int; let generatedAt, writerAppVersion, locale: String; let panchang: PanchangSlice; let verses: VerseSlice; let japam: JapamSlice }

enum PayloadState { case ready(WidgetPayload), missing, invalid, expired }

private let widgetIso: ISO8601DateFormatter = {
  let formatter = ISO8601DateFormatter(); formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]; return formatter
}()

func widgetDayFormatter(_ timeZone: String) -> DateFormatter {
  let formatter = DateFormatter(); formatter.calendar = Calendar(identifier: .gregorian); formatter.locale = Locale(identifier: "en_US_POSIX")
  formatter.timeZone = TimeZone(identifier: timeZone); formatter.dateFormat = "yyyy-MM-dd"; formatter.isLenient = false; return formatter
}

func widgetDateKey(_ date: Date, timeZone: String) -> String { widgetDayFormatter(timeZone).string(from: date) }

/// Payload instants are ISO-8601 with fractional seconds, the one format the
/// TypeScript writer emits and all three decoders agree on.
func widgetInstant(_ raw: String) -> Date? { widgetIso.date(from: raw) }

/// The tithi running at `at`, and its "तक …" line when this day's solve knows
/// the end — the same forward scan `runningTithi` performs in planner.ts and
/// `WidgetPayloadContract.runningTithi` performs in Kotlin. A payload written
/// before `tithiSegments` existed falls back to the day's sunrise tithi, so an
/// OTA'd JS writer and this binary never disagree about what is drawable.
func runningTithi(_ day: PanchangDay, at: Date) -> (name: LocalizedText, till: LocalizedText?) {
  guard let segments = day.tithiSegments, let last = segments.last else { return (day.tithi, nil) }
  let running = segments.first { segment in
    guard let raw = segment.endsAt, let ends = widgetInstant(raw) else { return true }
    return at <= ends
  } ?? last
  return (running.name, running.till)
}

private func validDateKey(_ key: String) -> Bool {
  guard key.range(of: #"^\d{4}-\d{2}-\d{2}$"#, options: .regularExpression) != nil,
        let date = widgetDayFormatter("UTC").date(from: key) else { return false }
  return widgetDayFormatter("UTC").string(from: date) == key
}

private func validLink(_ raw: String, path: String) -> Bool {
  guard let url = URLComponents(string: raw) else { return false }
  return url.scheme == "vedansh" && url.host == "widget" && url.path == path
}

private func exactLink(_ raw: String, path: String, query: [String: String]) -> Bool {
  guard validLink(raw, path: path), let items = URLComponents(string: raw)?.queryItems else { return query.isEmpty && URLComponents(string: raw)?.query == nil }
  guard items.count == query.count, Set(items.map(\.name)).count == items.count else { return false }
  return items.allSatisfy { item in item.value != nil && query[item.name] == item.value! }
}

/// Ordered, single-ended, anchored to the sunrise tithi — what `runningTithi`'s
/// single forward scan relies on. An unterminated MIDDLE link would stop that
/// scan early for every later instant, so it fails the payload closed.
private func validTithiSegments(_ segments: [PanchangTithi]?, sunriseTithi: LocalizedText) -> Bool {
  guard let segments = segments else { return true }
  guard let first = segments.first else { return false }
  var previous = Date.distantPast
  for (index, segment) in segments.enumerated() {
    guard let raw = segment.endsAt else {
      guard index == segments.count - 1, segment.till == nil else { return false }
      continue
    }
    guard let ends = widgetIso.date(from: raw), segment.till != nil, ends > previous else { return false }
    previous = ends
  }
  return widgetLanguages.allSatisfy { first.name.value($0) == sunriseTithi.value($0) }
}

private func validate(_ value: WidgetPayload) -> Bool {
  guard value.schemaVersion == widgetSupportedSchema, widgetIso.date(from: value.generatedAt) != nil,
        !value.writerAppVersion.isEmpty, widgetLanguages.contains(value.locale), value.panchang.timeZone == widgetIstTimeZone,
        TimeZone(identifier: value.verses.timeZone) != nil, TimeZone(identifier: value.japam.timeZone) != nil,
        !value.panchang.cityId.isEmpty, ["purnimant", "amanta"].contains(value.panchang.calendarSystem),
        widgetIso.date(from: value.panchang.validThrough) != nil, widgetIso.date(from: value.verses.validThrough) != nil,
        !value.panchang.days.isEmpty, !value.verses.days.isEmpty,
        Set(value.panchang.days.map(\.dateKey)).count == value.panchang.days.count,
        Set(value.verses.days.map(\.dateKey)).count == value.verses.days.count,
        validDateKey(value.japam.dateKey), value.japam.totalBeads >= 0, value.japam.totalRounds >= 0, value.japam.japaStreak >= 0,
        (value.japam.lastUsedMantraId == nil || !value.japam.lastUsedMantraId!.isEmpty),
        exactLink(value.japam.deepLink, path: "/japam", query: value.japam.lastUsedMantraId.map { ["mantraId": $0] } ?? [:]) else { return false }
  for day in value.panchang.days {
    guard validDateKey(day.dateKey), day.deepLink == "vedansh://widget/panchang?date=\(day.dateKey)",
          validTithiSegments(day.tithiSegments, sunriseTithi: day.tithi) else { return false }
  }
  for day in value.verses.days {
    guard validDateKey(day.dateKey), !day.sourceId.isEmpty, day.verseIndex >= 0, (day.chapter == nil || day.chapter! >= 1),
          [day.lines.hi, day.lines.en, day.lines.gu, day.lines.kn].allSatisfy({ !$0.isEmpty && $0.allSatisfy({ !$0.isEmpty }) }),
          exactLink(day.deepLink, path: "/verse", query: ["sourceId": day.sourceId, "verseIndex": String(day.verseIndex)].merging(day.chapter.map { ["chapter": String($0)] } ?? [:]) { current, _ in current }) else { return false }
  }
  return true
}

func decodeWidgetPayload(_ data: Data, now: Date = Date()) -> PayloadState {
  guard let value = try? JSONDecoder().decode(WidgetPayload.self, from: data), validate(value),
        let pValid = widgetIso.date(from: value.panchang.validThrough),
        let vValid = widgetIso.date(from: value.verses.validThrough) else { return .invalid }
  return pValid < now || vValid < now ? .expired : .ready(value)
}

func widgetPayloadExpiry(_ value: WidgetPayload) -> Date? {
  guard let pValid = widgetIso.date(from: value.panchang.validThrough),
        let vValid = widgetIso.date(from: value.verses.validThrough) else { return nil }
  return min(pValid, vValid)
}

File: mobile/plugins/home-widgets/android/WidgetPayloadContract.kt
package __APP_PACKAGE__.widgets

import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

internal object WidgetPayloadContract {
    sealed class State {
        data class Ready(val root: JSONObject) : State()
        data object Invalid : State()
        data object Expired : State()
    }

    private val languages = listOf("hi", "en", "gu", "kn")
    private val timeZones = TimeZone.getAvailableIDs().toSet()
    private val datePattern = Regex("^\\d{4}-\\d{2}-\\d{2}$")
    private val isoPattern = Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d{3}Z$")

    fun decode(raw: String, nowMs: Long = System.currentTimeMillis()): State {
        return try {
            val root = JSONObject(raw)
            if (integer(root, "schemaVersion") != 1 || parseIso(string(root, "generatedAt")) == null ||
                string(root, "writerAppVersion").isNullOrEmpty() || string(root, "locale") !in languages) return State.Invalid
            val panchang = objectValue(root, "panchang") ?: return State.Invalid
            val verses = objectValue(root, "verses") ?: return State.Invalid
            val japam = objectValue(root, "japam") ?: return State.Invalid
            if (string(panchang, "timeZone") != "Asia/Kolkata" ||
                !validTimeZone(string(verses, "timeZone")) || !validTimeZone(string(japam, "timeZone")) ||
                string(panchang, "cityId") == null || !localized(objectValue(panchang, "cityLabel")) ||
                string(panchang, "calendarSystem") !in listOf("purnimant", "amanta")) return State.Invalid
            val pValid = parseIso(string(panchang, "validThrough")) ?: return State.Invalid
            val vValid = parseIso(string(verses, "validThrough")) ?: return State.Invalid
            val pDays = arrayValue(panchang, "days") ?: return State.Invalid
            val vDays = arrayValue(verses, "days") ?: return State.Invalid
            if (!validDays(pDays, ::validPanchangDay) || !validDays(vDays, ::validVerseDay) || !validJapam(japam)) return State.Invalid
            if (pValid < nowMs || vValid < nowMs) State.Expired else State.Ready(root)
        } catch (_: Exception) { State.Invalid }
    }

    fun currentDateKey(timeZone: String, date: Date = Date()): String =
        SimpleDateFormat("yyyy-MM-dd", Locale.US).apply { this.timeZone = TimeZone.getTimeZone(timeZone) }.format(date)

    /**
     * The tithi running at [nowMs] and its "तक …" line — the same forward scan
     * `runningTithi` performs in planner.ts and VedanshWidgets.swift. A tithi
     * hands over MID-DAY, so a provider that drew `tithi` (the sunrise tithi)
     * for the whole civil day disagreed with the app's own glance from the
     * handover until midnight. Falls back to `tithi` for a payload written
     * before `tithiSegments` existed.
     */
    fun runningTithi(day: JSONObject, nowMs: Long): Pair<JSONObject, JSONObject?> {
        val segments = day.optJSONArray("tithiSegments")
        if (segments == null || segments.length() == 0) return day.getJSONObject("tithi") to null
        for (index in 0 until segments.length()) {
            val segment = segments.optJSONObject(index) ?: continue
            val ends = parseIso(segment.opt("endsAt") as? String)
            if (ends == null || nowMs <= ends) return segment.getJSONObject("name") to segment.optJSONObject("till")
        }
        val last = segments.getJSONObject(segments.length() - 1)
        return last.getJSONObject("name") to last.optJSONObject("till")
    }

    /**
     * The next instant this surface's drawn facts stop being true: the next
     * tithi handover for the Panchang, and in both cases the next civil-day
     * rollover in the slice's own zone.
     *
     * `updatePeriodMillis` alone cannot carry this. It is capped at 30 minutes,
     * it does NOT wake a dozing device, and a handover is a minute-level fact —
     * which is how a placed widget sat on yesterday's date through a morning
     * and on a finished tithi through an afternoon. The provider turns this
     * into one AlarmManager wake-up per boundary.
     */
    fun nextBoundaryMs(slice: JSONObject, timeZone: String, includeTithi: Boolean, nowMs: Long): Long {
        var next = nextMidnightMs(timeZone, nowMs)
        if (!includeTithi) return next
        val days = slice.optJSONArray("days") ?: return next
        for (index in 0 until days.length()) {
            val segments = days.optJSONObject(index)?.optJSONArray("tithiSegments") ?: continue
            for (link in 0 until segments.length()) {
                val ends = parseIso(segments.optJSONObject(link)?.opt("endsAt") as? String) ?: continue
                if (ends > nowMs && ends + 1_000L < next) next = ends + 1_000L
            }
        }
        return next
    }

    /** One second past the next midnight in [timeZone], so the new day's key is already current. */
    private fun nextMidnightMs(timeZone: String, nowMs: Long): Long {
        val calendar = Calendar.getInstance(TimeZone.getTimeZone(timeZone), Locale.US)
        calendar.timeInMillis = nowMs
        calendar.add(Calendar.DAY_OF_YEAR, 1)
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 1)
        calendar.set(Calendar.MILLISECOND, 0)
        return calendar.timeInMillis
    }

    private fun validDays(days: JSONArray, validator: (JSONObject) -> Boolean): Boolean {
        if (days.length() == 0) return false
        val keys = mutableSetOf<String>()
        for (index in 0 until days.length()) {
            val value = days.opt(index) as? JSONObject ?: return false
            val key = string(value, "dateKey") ?: return false
            if (!keys.add(key) || !validator(value)) return false
        }
        return true
    }

    private fun validPanchangDay(day: JSONObject): Boolean {
        val key = string(day, "dateKey") ?: return false
        return validDateKey(key) && localized(objectValue(day, "representedDate")) && localized(objectValue(day, "tithi")) &&
            (!day.has("vrat") || localized(objectValue(day, "vrat"))) && localized(objectValue(day, "sunrise")) &&
            validTithiSegments(day) &&
            localized(objectValue(day, "rahuKaal")) && (!day.has("abhijit") || localized(objectValue(day, "abhijit"))) &&
            string(day, "deepLink") == "vedansh://widget/panchang?date=$key"
    }

    /**
     * Ordered, single-ended, anchored to the sunrise tithi — what [runningTithi]'s
     * single forward scan relies on. An unterminated MIDDLE link would stop that
     * scan early for every later instant, so it fails the payload closed.
     */
    private fun validTithiSegments(day: JSONObject): Boolean {
        if (!day.has("tithiSegments")) return true
        val segments = arrayValue(day, "tithiSegments") ?: return false
        if (segments.length() == 0) return false
        var previous = Long.MIN_VALUE
        for (index in 0 until segments.length()) {
            val segment = segments.opt(index) as? JSONObject ?: return false
            if (!localized(objectValue(segment, "name"))) return false
            if (!segment.has("endsAt")) {
                if (index != segments.length() - 1 || segment.has("till")) return false
                continue
            }
            val ends = parseIso(string(segment, "endsAt")) ?: return false
            if (ends <= previous || !localized(objectValue(segment, "till"))) return false
            previous = ends
        }
        val first = objectValue(segments.getJSONObject(0), "name") ?: return false
        val sunrise = objectValue(day, "tithi") ?: return false
        return languages.all { string(first, it) == string(sunrise, it) }
    }

    private fun validVerseDay(day: JSONObject): Boolean {
        if (!validDateKey(string(day, "dateKey")) || string(day, "sourceId") == null ||
            (integer(day, "verseIndex") ?: -1) < 0 || (day.has("chapter") && (integer(day, "chapter") ?: 0) < 1)) return false
        val lines = objectValue(day, "lines") ?: return false
        if (!languages.all { lang ->
                val values = arrayValue(lines, lang) ?: return@all false
                values.length() > 0 && (0 until values.length()).all { values.opt(it) is String && (values.opt(it) as String).isNotEmpty() }
            }) return false
        val query = mutableMapOf("sourceId" to string(day, "sourceId")!!, "verseIndex" to integer(day, "verseIndex").toString())
        integer(day, "chapter")?.let { query["chapter"] = it.toString() }
        return localized(objectValue(day, "excerpt")) && localized(objectValue(day, "source")) &&
            localized(objectValue(day, "accessibilityLabel")) && exactLink(string(day, "deepLink"), "/verse", query)
    }

    private fun validJapam(japam: JSONObject): Boolean =
        validDateKey(string(japam, "dateKey")) && validTimeZone(string(japam, "timeZone")) &&
            listOf("totalBeads", "totalRounds", "japaStreak").all { (integer(japam, it) ?: -1) >= 0 } &&
            (!japam.has("lastUsedMantraId") || !string(japam, "lastUsedMantraId").isNullOrEmpty()) &&
            exactLink(string(japam, "deepLink"), "/japam", string(japam, "lastUsedMantraId")?.let { mapOf("mantraId" to it) } ?: emptyMap())

    private fun exactLink(raw: String?, path: String, expected: Map<String, String>): Boolean {
        if (raw == null) return false
        return try {
            val uri = android.net.Uri.parse(raw)
            if (uri.scheme != "vedansh" || uri.host != "widget" || uri.path != path || uri.fragment != null) return false
            val names = uri.queryParameterNames
            names.size == expected.size && names == expected.keys && names.all { uri.getQueryParameters(it).size == 1 && uri.getQueryParameter(it) == expected[it] }
        } catch (_: Exception) { false }
    }

    private fun localized(value: JSONObject?): Boolean = value != null && languages.all { string(value, it) != null }
    private fun validTimeZone(value: String?): Boolean = value != null && value in timeZones
    private fun validDateKey(value: String?): Boolean {
        if (value == null || !datePattern.matches(value)) return false
        return try { SimpleDateFormat("yyyy-MM-dd", Locale.US).apply { isLenient = false }.parse(value) != null } catch (_: Exception) { false }
    }
    private fun parseIso(value: String?): Long? {
        if (value == null || !isoPattern.matches(value)) return null
        return try { SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply { isLenient = false; timeZone = TimeZone.getTimeZone("UTC") }.parse(value)?.time } catch (_: Exception) { null }
    }
    private fun string(value: JSONObject, key: String): String? = value.opt(key) as? String
    private fun integer(value: JSONObject, key: String): Int? = (value.opt(key) as? Number)?.toInt()?.takeIf { number -> (value.opt(key) as Number).toDouble() == number.toDouble() }
    private fun objectValue(value: JSONObject, key: String): JSONObject? = value.opt(key) as? JSONObject
    private fun arrayValue(value: JSONObject, key: String): JSONArray? = value.opt(key) as? JSONArray
}
